nanoFramework.Device.Bluetooth/GenericAttributeProfile/GattLocalCharacteristic.cs

```diff
@@ -0,0 +1,400 @@
+        /// </summary>
+        public event GattLocalCharacteristicReadEventHandler ReadRequested;
+
+        /// <summary>
+        /// This is an event that is triggered when a write was requested.
+        /// </summary>
+        public event GattLocalCharacteristicWriteEventHandler WriteRequested;
+
+        /// <summary>
+        ///  An event that triggers when clients subscribed to this local characteristic changes.
+        /// </summary>
+        public event GattLocalCharacteristicClientsChangedEventHandler SubscribedClientsChanged;
+
+        internal void OnReadRequested(UInt16 descritorId, GattReadRequestedEventArgs e)
+        {
+            bool handled = false;
+
+            // Static value for Characteristic ?
+            if (_staticValue != null && descritorId == 0)
+            {
+                handled = true;
+                // Handle static values internally, don't fire an event
+                DataWriter writer = new DataWriter();
+                writer.WriteBuffer(_staticValue);
+                e.GetRequest().RespondWithValue(_staticValue);
+            }
+            else if (descritorId != 0)
+            {
+                // Descriptor event, let descriptor handle it
+                GattLocalDescriptor des = FindDescriptor(descritorId);
+                if (des != null)
+                {
+                    handled = des.OnReadRequested(e);
+                }
+            }
+            else if (ReadRequested != null)
+            {
+                // LocalCharacteristic event ?
+                handled = true;
+                ReadRequested?.Invoke(this, e);
+            }
+
+            if (!handled)
+            {
+                // No event handler, respond with error
+                e.GetRequest().RespondWithProtocolError(GattProtocolError.UnlikelyError);
+            }
+        }
+
+        internal void OnWriteRequested(UInt16 descritorId, GattWriteRequestedEventArgs e)
+        {
+            bool handled = false;
+
+            if (WriteRequested != null)
+            {
+                // LocalCharacteristic event ?
+                if (descritorId == 0)
+                {
+                    handled = true;
+                    WriteRequested?.Invoke(this, e);
+                }
+                else
+                {
+                    // Descriptor event
+                    GattLocalDescriptor des = FindDescriptor(descritorId);
+                    if (des != null)
+                    {
+                        handled = des.OnWriteRequested(e);
+                    }
+                }
+            }
+
+            if (!handled)
+            {
+                // No event handler, respond with error
+                e.GetRequest().RespondWithProtocolError(GattProtocolError.UnlikelyError);
+            }
+        }
+
+        private int ClientSubscribed(GattSubscribedClient client)
+        {
+            for(int index=0; index<_subscribedClients.Count; index++)
```

Comment:
a good ctr +k + d will fix the few formatting issues in the file. Just do it in all files, it will remove trailing spaces and more.

Reply:
don't you need to lock this one?

Reply:
All Ctrl + K , Ctrl +D.  Thought i had already done this but obviously miss a few

Reply:
Missing lock added